src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -231,51 +259,232 @@ def frobenius_charpoly(self, var='X'):
+                function ring
+
+        EXAMPLES::
+
+            sage: Fq = GF(25)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(6)
+            sage: phi = DrinfeldModule(A, [z, 4, 1, z])
+            sage: phi._frobenius_charpoly_gekeler('X')
+            X^3 + ((z2 + 2)*T^2 + (z2 + 2)*T + 4*z2 + 4)*X^2 + ... + (3*z2 + 2)*T^2 + (3*z2 + 3)*T + 4
+
+        ::
+
+            sage: Fq = GF(125)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(2)
+            sage: phi = DrinfeldModule(A, [z, 0, z])
+            sage: phi._frobenius_charpoly_gekeler('X')
+            Traceback (most recent call last):
+            NotImplementedError: 'Gekeler' algorithm failed
+
+        ALGORITHM:
 
-            See docstrings of methods :meth:`frobenius_norm` and
-            :meth:`frobenius_trace` for further details on the
-            computation of the norm and of the trace.
+        Construct a linear system based on the requirement that the
+        Frobenius satisfies a degree r polynomial with coefficients in
+        the function ring. This generalizes the procedure from
+        [Gek2008]_ for the rank 2 case.
         """
-        self._check_rank_two()
-        A = self._function_ring  # Fq[T]
-        S = PolynomialRing(A, name=var)  # Fq[T][X]
-        # Does not work when Fq is not a prime field...
-        # chi = self._gen.reduced_charpoly()
-        # return -chi(A.gen(), S.gen())
-        return S([self.frobenius_norm(), -self.frobenius_trace(), 1])
+        K = self.base_over_constants_field()
+        A = self.function_ring()
+        r, n = self.rank(), self._base_degree_over_constants
+        # Compute constants that determine the block structure of the
+        # linear system. The system is prepared such that the solution
+        # vector has the form [a_0, a_1, ... a_{r-1}]^T with each a_i
+        # corresponding to a block of length (n*(r - i))//r + 1
+        shifts = [(n*(r - i))//r + 1 for i in range(r)]
+        rows, cols = n*r + 1, sum(shifts)
+        block_shifts = [0]
+        for i in range(r-1):
+            block_shifts.append(block_shifts[-1] + shifts[i])
+        # Compute the images \phi_T^i for i = 0 .. n.
+        gen_powers = [self(A.gen()**i).coefficients(sparse=False)
+                      for i in range(0, n + 1)]
+        sys, vec = Matrix(K, rows, cols), vector(K, rows)
+        vec[rows - 1] = -1
+        for j in range(r):
+            for k in range(shifts[j]):
+                for i in range(len(gen_powers[k])):
+                    sys[i + n*j, block_shifts[j] + k] = gen_powers[k][i]
+        if sys.right_nullity() != 0:
+            raise NotImplementedError("'Gekeler' algorithm failed")
+        sol = list(sys.solve_right(vec))
+        # The system is solved over L, but the coefficients should all be in Fq
+        # We project back into Fq here.
+        sol_Fq = list(map(lambda x: K(x).vector()[0], sol))
```

Comment:
```suggestion
        sol_Fq = [K(x).vector()[0] for x in sol]
```
plus same remark as before: this should be the work of `RingExtension`.